src/rigid_pendulum_poc/heavy_top.cpp

```diff
@@ -1,88 +1,44 @@
+        heavy_top_constraint_gradient_matrix(position_vector, rotation_matrix);
+
+    auto first_term = multiply_matrix_with_vector(mass_matrix, acceleration_vector);
+    auto second_term = gen_forces_vector;
+    auto third_term = multiply_matrix_with_vector(
+        transpose_matrix(constraint_gradient_matrix), lagrange_multipliers
     );
 
-    // Only the 3 x 3 lower right block of the tangent damping matrix is non-zero
-    auto tangent_damping_matrix = HostView2D("tangent_damping_matrix", 6, 6);
+    auto residual_vector = HostView1D("residual_vector", 6);
     Kokkos::parallel_for(
-        6,
-        KOKKOS_LAMBDA(const int i) {
-            for (size_t j = 0; j < 6; j++) {
-                if (i < 3 && j < 3) {
-                    tangent_damping_matrix(i, j) = 0.;
-                } else if (i < 3 && j >= 3) {
-                    tangent_damping_matrix(i, j) = 0.;
-                } else if (i >= 3 && j < 3) {
-                    tangent_damping_matrix(i, j) = 0.;
-                } else {
-                    tangent_damping_matrix(i, j) = nonzero_block(i - 3, j - 3);
-                }
-            }
-        }
+        6, KOKKOS_LAMBDA(const int i
+           ) { residual_vector(i) = first_term(i) + second_term(i) + third_term(i); }
     );
 
-    return tangent_damping_matrix;
-}
-
-HostView2D heavy_top_tangent_stiffness_matrix(
-    HostView1D position_vector, HostView2D rotation_matrix, HostView1D lagrange_multipliers
-) {
-    // Tangent stiffness matrix for the heavy top problem is given by
-    // [K_t] = [ [0]_3x3              [0]_3x3
-    //           [0]_3x3    [ ~{X} * ~([R^T] * {Lambda}) ] ]
-    auto X = create_cross_product_matrix(position_vector);
-
-    auto RT_Lambda =
-        multiply_matrix_with_vector(transpose_matrix(rotation_matrix), lagrange_multipliers);
-    auto RT_Lambda_matrix = create_cross_product_matrix(RT_Lambda);
-
-    auto non_zero_block = multiply_matrix_with_matrix(X, RT_Lambda_matrix);
-
-    // Only the 3 x 3 lower right block of the tangent stiffness matrix is non-zero
-    auto tangent_stiffness_matrix = HostView2D("tangent_stiffness_matrix", 6, 6);
-    Kokkos::parallel_for(
-        6,
-        KOKKOS_LAMBDA(const int i) {
-            for (size_t j = 0; j < 6; j++) {
-                if (i < 3 && j < 3) {
-                    tangent_stiffness_matrix(i, j) = 0.;
-                } else if (i < 3 && j >= 3) {
-                    tangent_stiffness_matrix(i, j) = 0.;
-                } else if (i >= 3 && j < 3) {
-                    tangent_stiffness_matrix(i, j) = 0.;
-                } else {
-                    tangent_stiffness_matrix(i, j) = non_zero_block(i - 3, j - 3);
-                }
-            }
-        }
-    );
+    auto log = util::Log::Get();
+    log->Debug("Residual vector is " + std::to_string(6) + " x 1 with elements\n");
+    for (size_t i = 0; i < 6; i++) {
+        log->Debug(std::to_string(residual_vector(i)) + "\n");
+    }
```

Comment:
I think this looks good in general. We may need to make some adjustments (probably just add a deep-copy to bring residual_vector back to host) once we start getting things to run on GPU. Maybe we would only do that for debug-mode, if logging is only active in debug-mode, or something like that.

Reply:
@alanw0  I am running into a peculiar memory related issue as follows. I wonder if it has something to do with copying Kokkos::Views as you describe above. Do you have any suggestions as to how to navigate this or expand on the above?

```
openturbine_unit_tests(43306,0x1ff149e00) malloc: Heap corruption detected, free list is damaged at 0x600003644620
*** Incorrect guard value: 0
openturbine_unit_tests(43306,0x1ff149e00) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

Reply:
If you are not running on GPU, you shouldn't run into the issue I was anticipating above.
Assuming you are running in a pure CPU build, you'll probably need to run this through valgrind or asan to see where the memory error is coming from. If you don't have access to one of those tools, we can do that for you and help figure it out.

Reply:
I am running on CPU only and not completely sure about the source of the error here. It would be really helpful if you could take a look and I will also keep investigating myself - this appears to be the only remaining hurdle in completing the heavy top problem.

I am pushing the remaining local changes from my end - please try to run the `TEST(HeavyTopProblemFromBrulsAndCardona2010PaperTest, AlphaStepSolutionAfterOneInc)` unit test. It appears fpr this particular test with high value for the `double final_time{0.2}` variable.

Reply:
ok I'll see what I can figure out.

Reply:
valgrind says there is an 'invalid write', which usually means writing past the end of an array allocation, at line 108 of `generalized_alpha_time_integrator.cpp`. That line is `lagrange_mults_next(i) = 0.;`. It is in a Kokkos::parallel_for which is looping over `velocity.size()`. So it is assuming that velocity and lagrange_mults_next are the same length, and they actually are different lengths?

Reply:
They are not the same size, this was exactly what the issue was. Thanks for your help, greatly appreciated!

Reply:
valgrind also says there is a bad read at line 164 of utilities.cpp, which is `sum += matrix_a(row, i) * matrix_b(i, column);`.
Again this hints at an array-bounds problem, but it isn't obvious to me what the problem is. It looks like you've got the sizes correct, and you've made sure that n_rows == n_columns, so it seems like it should be ok...

Reply:
Ok I will look into it.

Reply:
I think I see it now, on line 158/159 where you set the 'entries' object, it should have bounds of A_n_rows by B_n_columns.

Reply:
Yup, I see that as well.